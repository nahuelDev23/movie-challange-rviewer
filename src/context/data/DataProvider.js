import React, { useReducer } from 'react'
import api from '../../data'
import { DataContext } from './DataContext'
import { dataReducer } from './dataReducer'

const INITIAL_DATA_STATE = {
  series: [],
  movies: []
}

export const DataProvider = ({ children }) => {
  const [state, dispatch] = useReducer(dataReducer, INITIAL_DATA_STATE)

  const onLoadData = async (typeData) => {

    const data = await api.list()

    const filteredSeries = data[0].entries.filter(serie => serie.programType === typeData)

    const typeForDispatch = typeData === 'series' ? 'SET_SERIES' : 'SET_MOVIES'

    dispatch({
      type: typeForDispatch,
      payload: [...filteredSeries]

    })

  }


  return (
    <DataContext.Provider value={{
      ...state,

      //methods
      onLoadData

    }}>
      {children}
    </DataContext.Provider>
  )
}
