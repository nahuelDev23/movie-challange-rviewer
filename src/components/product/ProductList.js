import React, { useEffect, useContext } from 'react'
import { DataContext } from '../../context/data/DataContext'
import { ProductCard } from './ProductCard'
import { usePaginate } from '../../hooks/usePaginate'

export const ProductList = ({ type }) => {

  const { movies, series, onLoadData } = useContext(DataContext)

  const data = type === 'movie' ? movies : series

  const { currentData, currentPage, totalPages, paginate } = usePaginate(data, 10)

  const pages = Array.from({ length: totalPages }, (_, i) => i + 1)

  useEffect(() => {
    onLoadData(type)
  }, [])

  return (
    <div className='product'>
      <div className="product__container">
        {
          currentData && currentData.map(serie => (
            <ProductCard key={serie.title} product={serie} />
          ))
        }
      </div>
      {
        totalPages > 1 && (
          <div className="product__pagination">
            <button className="product__pagination__prev" onClick={() => paginate(currentPage - 1)}> </button>
            <ul className="product__pagination__list">
              {
                pages.map(page => (
                  <li onClick={() => paginate(page)} className={`product__pagination__item ${currentPage === page ? 'item_active' : undefined} `}>{page}</li>
                ))
              }
            </ul>

            <button className="product__pagination__next" onClick={() => paginate(currentPage + 1)}></button>
          </div>
        )
      }
    </div>
  )
}
