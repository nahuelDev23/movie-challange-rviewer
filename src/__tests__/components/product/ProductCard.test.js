import { mount } from 'enzyme'
import { ProductCard } from '../../../components/product/ProductCard';
import { DataContext } from '../../../context/data/DataContext';
import { UiContext } from '../../../context/ui/UiContext';


const data = {
    "title": "No Activity",
    "description": "Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat.",
    "programType": "series",
    "images": {
        "Poster Art": {
            "url": "https://streamcoimg-a.akamaihd.net/000/106/36/10636-PosterArt-9add943c5e62c2d89f6d31458d33508a.jpg",
            "width": 1000,
            "height": 1500
        }
    },
    "releaseYear": 2015
}
describe('Pruebas en <ProductCard/>', () => {

    test('Debe mostrar <ProductCard/> correctamente', () => {

        const wrapper = mount(
            <DataContext.Provider value={{
                onFilterDataByNameAndCategory: jest.fn(),
                onLoadData: jest.fn(),
                movies: [{}],
                series: [{}],
            }}>
                <UiContext.Provider value={{
                    isOpenFilter: true
                }}>
                    <ProductCard product={data} />
                </UiContext.Provider>
            </DataContext.Provider>

        );
        expect(wrapper).toMatchSnapshot();
    })

    test('Debe verse en pantalla  los mismos datos que hay en el objeto', () => {

        const wrapper = mount(
            <DataContext.Provider value={{
                onFilterDataByNameAndCategory: jest.fn(),
                onLoadData: jest.fn(),
                movies: [{}],
                series: [{}],
            }}>
                <UiContext.Provider value={{
                    isOpenFilter: true
                }}>
                    <ProductCard product={data} />
                </UiContext.Provider>
            </DataContext.Provider>

        );

        expect(wrapper.find('h1').text()).toBe(data.title)
        expect(wrapper.find('h2').text()).toBe(data.releaseYear.toString())
        expect(wrapper.find('p').text()).toBe(data.description.slice(0, 100))
        expect(wrapper.find('img').props().src).toBe(data.images['Poster Art'].url)


    })



})