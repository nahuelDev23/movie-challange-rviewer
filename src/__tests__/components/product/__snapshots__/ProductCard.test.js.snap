// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <ProductCard/> Debe mostrar <ProductCard/> correctamente 1`] = `
<div
  className="product__item"
  key="No Activity"
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
>
  <div
    className="product__title undefined"
  >
    No Activity
  </div>
  <div
    className="product__summary undefined"
  >
    <h1
      className="product__summary_title"
    >
      No Activity
    </h1>
    <h2
      className="product__summary_subtitle"
    >
      2015
    </h2>
    <p
      className="product__summary_description"
    >
      Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore 
    </p>
  </div>
  <img
    alt="poster No Activity"
    className="product__img"
    src="https://streamcoimg-a.akamaihd.net/000/106/36/10636-PosterArt-9add943c5e62c2d89f6d31458d33508a.jpg"
  />
</div>
`;
